buffer/src/main/java/io/netty/buffer/LeakAwareByteBufUtil.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer;
+
+import io.netty.util.ResourceLeak;
+
+final class LeakAwareByteBufUtil {
+
+    private LeakAwareByteBufUtil() { }
+
+    static ByteBuf unwrappedDerived(ByteBuf leakAware, ResourceLeak parentLeak, ByteBuf derived) {
+        ByteBuf unwrappedDerived = unwrapSwapped(derived);
+
+        if (unwrappedDerived instanceof AbstractPooledDerivedByteBuf) {
+            ResourceLeak leak = AbstractByteBuf.leakDetector.open(derived);
+            if (leak == null) {
+                leak = new NoopResourceLeak();
```

Comment:
can we just return `derived` in this case? What benefit does wrapping in a "leak aware buffer" with `NoopResourceLeak` provide? Or should we add a package private method on `ResourceLeakDetector` so that we can force a `DefaultResourceLeak` to be allocated (only used in this situation when the buffer is derived from another buffer which already has leak info)?